src/content/app-architecture/case-study/data-layer.md

```diff
@@ -0,0 +1,322 @@
+* **Services** are stateless Dart classes that interact with APIs, like HTTP
+  servers and platform plugins. Any data that your application needs that isn't
+  created inside the application code itself should be fetched from within
+  service classes.
+
+## Define a Service
+
+A service class is the least ambiguous of all the architecture components. 
+It's stateless, and its functions don't have side effects. 
+Its only job is to wrap an external API. 
+There is generally one service class per data source, 
+such as a client HTTP server or a platform plugin.
+
+
+
+![A diagram that shows the inputs and outputs of service objects.](/assets/images/docs/app-architecture/case-study/mvvm-case-study-services-architecture.png)
+
+In the Compass app, for example, there is an [`APIClient`][] service that 
+handles the CRUD calls to the client-facing server.
+
+```dart title=api_client.dart
+class ApiClient {
+  // ... some code omitted for demo purposes
+
+  Future<Result<List<ContinentApiModel>>> getContinents() async {...}
+
+  Future<Result<List<DestinationApiModel>>> getDestinations() async {...}
+
+  Future<Result<List<ActivityApiModel>>> getActivityByDestination(String ref) async {...}
+
+  Future<Result<List<BookingApiModel>>> getBookings() async {...}
+
+  Future<Result<BookingApiModel>> getBooking(int id) async {...}
+
+  Future<Result<BookingApiModel>> postBooking(BookingApiModel booking) async {...}
+
+  Future<Result<void>> deleteBooking(int id) async {...}
+
+  Future<Result<UserApiModel>> getUser() async {...}
+}
+```
+
+The service itself is a class, 
+where each method wraps a different API endpoint and 
+exposes asynchronous response objects. 
+Continuing the earlier example of deleting a saved booking, 
+the `deleteBooking` method returns a `Future<Result<void>>`.
+
+:::note
+Some methods return data classes that are 
+specifically for raw data from the API, 
+such as the `BookingApiModel` class. 
+As you'll soon see, repositories extract data and 
+expose it to `ViewModel`s in a different format.
+:::
+
+### Result objects
+
+In this app, methods on service classes return `Result` objects.
+`Result` is a utility class that wraps asynchronous calls,  
+making it easier to handle errors. 
+Encoding asynchronous responses in this way makes it easy to
+write helper functions for a number of different cases, 
+such as API retry logic, 
+without having to write custom logic for every instance of the case.
+
+In the Compass app, 
+`Result` is a [`sealed`][] class with two subclasses, 
+called `Ok` and `Error`. 
+The following example shows the `Result` class.
+
+```dart title=result.dart
+sealed class Result<T> {
+  const Result();
+
+  /// Creates an instance of [Result] containing a value.
+  factory Result.ok(T value) => Ok(value);
+
+  /// Creates an instance of [Result] containing an error.
+  factory Result.error(Exception error) => Error(error);
+  
+  Ok<T> get asOk => this as Ok<T>;
+  
+  Error get asError => this as Error<T>;
```

Comment:
We had a long conversation about this while implementing the app, and landed on this being a valuable tool for a complex app. 

That said, I'm now in favor of removing this section from this page because it duplicates content from #11444. I agree that this pattern isn't _necessary_, so it's probably better presented in the cookbook recipe.

I replaced this section with a short note and link.